### PR TITLE
curvebs/mds: add cluster io metrics

### PR DESCRIPTION
**Pull request by wuhongsong**
signed-off-by: hzwuhongsong hzwuhongsong@corp.netease.com

### CODE REVIEW

**Comment by YunhuiChen** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
image


image

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

File: src/mds/topology/topology_metric.h [hunk 1]
```diff
@@ -238,8 +238,34 @@ struct LogicalPoolMetric {
...
 using LogicalPoolMetricPtr = std::unique_ptr<LogicalPoolMetric>;
 
+struct ClusterMetric {
+    const std::string kTopologyClusterMetricPrefix =
+        "topology_metric_cluster_";
+
+    // bandwidth of reading
+    bvar::Status<uint32_t> readRate;
+    // bandwidth of writing
+    bvar::Status<uint32_t> writeRate;
```
Comment (aspirer) [inline src/mds/topology/topology_metric.h:248]: 2<<31
4294967296
略等于4G/s，应该不够用。

**Review by aspirer — CHANGES_REQUESTED** [review]

File: src/mds/topology/topology_metric.cpp [hunk 2]
```diff
@@ -330,6 +355,7 @@ void TopologyMetricService::BackEndFunc() {
 
 int TopologyMetricService::Init(const TopologyOption &option) {
     option_ = option;
+
```
Comment (aspirer) [inline src/mds/topology/topology_metric.cpp:358]: delete this blank line?

**Review by aspirer — CHANGES_REQUESTED** [review]

**Comment by wuhongsong** [comment]
cicheck

**Comment by wuhongsong** [comment]
cicheck

File: src/mds/topology/topology_metric.cpp [hunk 3]
```diff
@@ -232,6 +234,29 @@ void TopologyMetricService::UpdateTopologyMetrics() {
         }
     }
 
+    // cluster io
+    uint64_t clusterReadIOPS = 0, clusterWriteIOPS = 0,
+        clusterReadRate = 0, clusterWriteRate = 0;
+    for (auto metricIter = gChunkServerMetrics.begin();
```
Comment (SeanHai) [inline src/mds/topology/topology_metric.cpp:240]: The cluster read/write iops and rate only need add all the pools' related metric value( https://github.com/opencurve/curve/pull/2176)

File: src/mds/topology/topology_metric.cpp [hunk 4]
```diff
@@ -232,6 +234,29 @@ void TopologyMetricService::UpdateTopologyMetrics() {
         }
     }
 
+    // cluster io
+    uint64_t clusterReadIOPS = 0, clusterWriteIOPS = 0,
```
Comment (SeanHai) [inline src/mds/topology/topology_metric.cpp:238]: It isbettrer to keep the data type consistent with metric

File: src/mds/topology/topology_metric.h [hunk 5]
```diff
@@ -238,8 +238,34 @@ struct LogicalPoolMetric {
 };
 using LogicalPoolMetricPtr = std::unique_ptr<LogicalPoolMetric>;
 
+struct ClusterMetric {
```
Comment (SeanHai) [inline src/mds/topology/topology_metric.h:241]: The poolNum, serverNum and chunkserverNum of cluster is also needed.

File: src/mds/topology/topology_metric.h (hunk 1 above)
Comment (wuhongsong) [reply src/mds/topology/topology_metric.h]: fxi

File: src/mds/topology/topology_metric.cpp (hunk 2 above)
Comment (wuhongsong) [reply src/mds/topology/topology_metric.cpp]: fix

File: src/mds/topology/topology_metric.cpp (hunk 4 above)
Comment (wuhongsong) [reply src/mds/topology/topology_metric.cpp:273]: fix

File: src/mds/topology/topology_metric.h (hunk 5 above)
Comment (wuhongsong) [reply src/mds/topology/topology_metric.h:259]: fix

**Comment by wuhongsong** [comment]
cicheck

File: src/mds/topology/topology_metric.cpp (hunk 3 above)
Comment (wuhongsong) [reply src/mds/topology/topology_metric.cpp]: fix

File: src/mds/topology/topology_metric.h (hunk 5 above)
Comment (SeanHai) [reply src/mds/topology/topology_metric.h:259]: The serverNum is missed.

**Review by aspirer — APPROVED** [review]

**Comment by wuhongsong** [comment]
cicheck

**Review by SeanHai — APPROVED** [review]

**Pull request merged**